File: app_mobile/kmp/androidApp/src/main/java/com/ansmi/gestsquadre/kmp/map/MapMarkerFactory.kt
package com.ansmi.gestsquadre.kmp.map

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.graphics.RectF
import android.graphics.Typeface
import android.graphics.drawable.BitmapDrawable
import androidx.core.content.ContextCompat
import com.ansmi.gestsquadre.shared.model.LiveSquadPin
import com.ansmi.gestsquadre.shared.model.MapWaypointPin
import org.osmdroid.views.overlay.Marker
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

private object MapMarkerMetrics {
    fun dp(context: Context, value: Float): Float =
        value * context.resources.displayMetrics.density

    fun sp(context: Context, value: Float): Float =
        value * context.resources.displayMetrics.scaledDensity
}

private const val SQUAD_CHIP_MAX_CHARS = 28

private fun squadMapChipLabel(squad: LiveSquadPin, alarming: Boolean): String {
    val base =
        squad.squadName
            .trim()
            .ifEmpty { squad.squadCode }
            .take(SQUAD_CHIP_MAX_CHARS)
            .uppercase()
    return if (alarming) "⚠ $base" else base
}

object MapMarkerFactory {
    fun squadMarker(
        context: Context,
        squad: LiveSquadPin,
        alarming: Boolean,
        isSelf: Boolean,
    ): Bitmap {
        val iconHeightPx =
            MapMarkerMetrics.dp(context, when {
                isSelf -> 32f
                alarming -> 30f
                else -> 28f
            }).roundToInt()
        val iconBitmap = loadIconBitmap(context, squad.mapIconKey, iconHeightPx, SquadIcons::drawableRes)
        val label = squadMapChipLabel(squad, alarming)
        val labelBg = if (alarming) 0xFFC62828.toInt() else 0xFF111111.toInt()
        val labelBorder = if (alarming) Color.WHITE else 0xFF4A5568.toInt()

        val textPaint =
            Paint(Paint.ANTI_ALIAS_FLAG).apply {
                color = Color.WHITE
                textSize = MapMarkerMetrics.sp(context, 10f)
                typeface = Typeface.create(Typeface.DEFAULT, Typeface.BOLD)
                textAlign = Paint.Align.CENTER
            }
        val textWidth = textPaint.measureText(label)
        val width =
            max(
                max(iconBitmap.width.toFloat(), textWidth) + MapMarkerMetrics.dp(context, 18f),
                MapMarkerMetrics.dp(context, 168f),
            ).roundToInt()
        val labelHeight = MapMarkerMetrics.dp(context, 16f).roundToInt()
        val height = iconBitmap.height + labelHeight + MapMarkerMetrics.dp(context, 4f).roundToInt()
        val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        bitmap.eraseColor(Color.TRANSPARENT)
        val canvas = Canvas(bitmap)
        val cx = width / 2f

        canvas.drawBitmap(
            iconBitmap,
            (cx - iconBitmap.width / 2f).roundToInt().toFloat(),
            MapMarkerMetrics.dp(context, 1f),
            null,
        )

        val labelTop = iconBitmap.height + MapMarkerMetrics.dp(context, 3f)
        val labelRect =
            RectF(
                MapMarkerMetrics.dp(context, 8f),
                labelTop,
                width - MapMarkerMetrics.dp(context, 8f),
                height - MapMarkerMetrics.dp(context, 2f),
            )
        val labelPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { color = labelBg }
        val labelBorderPaint =
            Paint(Paint.ANTI_ALIAS_FLAG).apply {
                color = labelBorder
                style = Paint.Style.STROKE
                strokeWidth = MapMarkerMetrics.dp(context, if (alarming) 2f else 1.5f)
            }
        val corner = MapMarkerMetrics.dp(context, 6f)
        canvas.drawRoundRect(labelRect, corner, corner, labelPaint)
        canvas.drawRoundRect(labelRect, corner, corner, labelBorderPaint)
        canvas.drawText(
            label,
            cx,
            labelTop + MapMarkerMetrics.sp(context, 11f),
            textPaint,
        )
        return bitmap
    }

    fun waypointMarker(
        context: Context,
        waypoint: MapWaypointPin,
    ): Bitmap {
        // Allineato a Flutter (26px) / TOC web (28px icona, chip 9px).
        val iconHeightPx = MapMarkerMetrics.dp(context, 28f).roundToInt()
        val iconBitmap = loadIconBitmap(context, waypoint.iconKey, iconHeightPx, WaypointIcons::drawableRes)
        val label = waypoint.displayName.uppercase()
        val textPaint =
            Paint(Paint.ANTI_ALIAS_FLAG).apply {
                color = Color.BLACK
                textSize = MapMarkerMetrics.sp(context, 9f)
                typeface = Typeface.create(Typeface.DEFAULT, Typeface.BOLD)
                textAlign = Paint.Align.CENTER
            }
        val textWidth = textPaint.measureText(label)
        val width =
            max(
                max(iconBitmap.width.toFloat(), textWidth) + MapMarkerMetrics.dp(context, 14f),
                MapMarkerMetrics.dp(context, 72f),
            ).roundToInt()
        val labelHeight = MapMarkerMetrics.dp(context, 14f).roundToInt()
        val height = iconBitmap.height + labelHeight + MapMarkerMetrics.dp(context, 2f).roundToInt()
        val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        bitmap.eraseColor(Color.TRANSPARENT)
        val canvas = Canvas(bitmap)
        val cx = width / 2f

        canvas.drawBitmap(
            iconBitmap,
            cx - iconBitmap.width / 2f,
            0f,
            null,
        )

        val labelTop = iconBitmap.height + MapMarkerMetrics.dp(context, 1f)
        val padH = MapMarkerMetrics.dp(context, 5f)
        val labelRect =
            RectF(
                max(MapMarkerMetrics.dp(context, 4f), cx - textWidth / 2f - padH),
                labelTop,
                min(width - MapMarkerMetrics.dp(context, 4f), cx + textWidth / 2f + padH),
                height - MapMarkerMetrics.dp(context, 1f),
            )
        val labelPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { color = 0xFFFF6969.toInt() }
        val labelBorder =
            Paint(Paint.ANTI_ALIAS_FLAG).apply {
                color = Color.WHITE
                style = Paint.Style.STROKE
                strokeWidth = MapMarkerMetrics.dp(context, 1.2f)
            }
        val corner = MapMarkerMetrics.dp(context, 5f)
        canvas.drawRoundRect(labelRect, corner, corner, labelPaint)
        canvas.drawRoundRect(labelRect, corner, corner, labelBorder)
        canvas.drawText(
            label,
            cx,
            labelTop + MapMarkerMetrics.sp(context, 9f),
            textPaint,
        )
        return bitmap
    }

    private fun loadIconBitmap(
        context: Context,
        iconKey: String?,
        targetHeightPx: Int,
        drawableRes: (String?) -> Int,
    ): Bitmap {
        val drawable =
            ContextCompat.getDrawable(context, drawableRes(iconKey))
                ?: return Bitmap.createBitmap(targetHeightPx, targetHeightPx, Bitmap.Config.ARGB_8888)

        val srcW = drawable.intrinsicWidth.coerceAtLeast(1)
        val srcH = drawable.intrinsicHeight.coerceAtLeast(1)
        val scale = targetHeightPx.toFloat() / srcH.toFloat()
        val outW = max((srcW * scale).roundToInt(), 1)
        val outH = max(targetHeightPx, 1)
        val bitmap = Bitmap.createBitmap(outW, outH, Bitmap.Config.ARGB_8888)
        bitmap.eraseColor(Color.TRANSPARENT)
        val canvas = Canvas(bitmap)
        drawable.bounds = Rect(0, 0, outW, outH)
        drawable.draw(canvas)
        return bitmap
    }

    fun applyMarkerIcon(
        context: Context,
        marker: Marker,
        bitmap: Bitmap,
    ) {
        val drawable = BitmapDrawable(context.resources, bitmap)
        drawable.setBounds(0, 0, bitmap.width, bitmap.height)
        drawable.setAntiAlias(true)
        drawable.isFilterBitmap = true
        marker.icon = drawable
        marker.setAnchor(Marker.ANCHOR_CENTER, Marker.ANCHOR_BOTTOM)
        marker.setInfoWindow(null)
    }
}

File: app_mobile/kmp/iosApp/iosApp/MapMarkerFactory.swift
import UIKit
import shared

private let squadChipMaxChars = 28

enum MapMarkerFactory {
    static func squadMarker(
        squad: LiveSquadPin,
        alarming: Bool,
        isSelf: Bool,
        scale: CGFloat = UIScreen.main.scale
    ) -> UIImage {
        let density = scale / UIScreen.main.scale
        let iconHeight: CGFloat = {
            if isSelf { return 32 * density }
            if alarming { return 30 * density }
            return 28 * density
        }()
        let iconBitmap = loadIconBitmap(
            name: SquadIcons.bundleImageName(forKey: squad.mapIconKey),
            targetHeight: iconHeight
        )
        let label = squadMapChipLabel(squad: squad, alarming: alarming)
        let labelBg = alarming ? UIColor(red: 0.78, green: 0.16, blue: 0.16, alpha: 1) : UIColor(white: 0.07, alpha: 1)
        let labelBorder = alarming ? UIColor.white : UIColor(red: 0.29, green: 0.33, blue: 0.41, alpha: 1)

        let textFont = UIFont.boldSystemFont(ofSize: 10 * density)
        let textAttrs: [NSAttributedString.Key: Any] = [.font: textFont, .foregroundColor: UIColor.white]
        let textWidth = (label as NSString).size(withAttributes: textAttrs).width
        let width = max(max(iconBitmap.size.width, textWidth) + 18 * density, 168 * density)
        let labelHeight = 16 * density
        let height = iconBitmap.size.height + labelHeight + 4 * density

        let renderer = UIGraphicsImageRenderer(size: CGSize(width: width, height: height))
        return renderer.image { _ in
            let cx = width / 2
            iconBitmap.draw(at: CGPoint(x: cx - iconBitmap.size.width / 2, y: 1 * density))

            let labelTop = iconBitmap.size.height + 3 * density
            let labelRect = CGRect(
                x: 8 * density,
                y: labelTop,
                width: width - 16 * density,
                height: height - labelTop - 2 * density
            )
            let corner = 6 * density
            let path = UIBezierPath(roundedRect: labelRect, cornerRadius: corner)
            labelBg.setFill()
            path.fill()
            labelBorder.setStroke()
            path.lineWidth = (alarming ? 2 : 1.5) * density
            path.stroke()

            drawCenteredLabel(label, in: labelRect, attributes: textAttrs)
        }
    }

    static func waypointMarker(waypoint: MapWaypointPin, scale: CGFloat = UIScreen.main.scale) -> UIImage {
        let density = scale / UIScreen.main.scale
        let iconHeight = 28 * density
        let iconBitmap = loadIconBitmap(name: WaypointIcons.imageName(for: waypoint.iconKey), targetHeight: iconHeight)
        let label = waypoint.displayName.uppercased()
        let textFont = UIFont.boldSystemFont(ofSize: 9 * density)
        let textAttrs: [NSAttributedString.Key: Any] = [.font: textFont, .foregroundColor: UIColor.black]
        let textWidth = (label as NSString).size(withAttributes: textAttrs).width
        let width = max(max(iconBitmap.size.width, textWidth) + 14 * density, 72 * density)
        let labelHeight = 14 * density
        let height = iconBitmap.size.height + labelHeight + 2 * density

        let renderer = UIGraphicsImageRenderer(size: CGSize(width: width, height: height))
        return renderer.image { _ in
            let cx = width / 2
            iconBitmap.draw(at: CGPoint(x: cx - iconBitmap.size.width / 2, y: 0))

            let labelTop = iconBitmap.size.height + 1 * density
            let padH = 5 * density
            let labelRect = CGRect(
                x: max(4 * density, cx - textWidth / 2 - padH),
                y: labelTop,
                width: min(width - 8 * density, textWidth + padH * 2),
                height: height - labelTop - 1 * density
            )
            let path = UIBezierPath(roundedRect: labelRect, cornerRadius: 5 * density)
            UIColor(red: 1, green: 0.41, blue: 0.41, alpha: 1).setFill()
            path.fill()
            UIColor.white.setStroke()
            path.lineWidth = 1.2 * density
            path.stroke()
            drawCenteredLabel(label, in: labelRect, attributes: textAttrs)
        }
    }

    private static func drawCenteredLabel(
        _ text: String,
        in rect: CGRect,
        attributes: [NSAttributedString.Key: Any]
    ) {
        let paragraph = NSMutableParagraphStyle()
        paragraph.alignment = .center
        var attrs = attributes
        attrs[.paragraphStyle] = paragraph
        let bounding = (text as NSString).boundingRect(
            with: CGSize(width: rect.width, height: .greatestFiniteMagnitude),
            options: [.usesLineFragmentOrigin, .usesFontLeading],
            attributes: attrs,
            context: nil
        )
        let drawRect = CGRect(
            x: rect.minX,
            y: rect.minY + max(0, (rect.height - bounding.height) / 2),
            width: rect.width,
            height: bounding.height
        )
        (text as NSString).draw(
            with: drawRect,
            options: [.usesLineFragmentOrigin, .usesFontLeading],
            attributes: attrs,
            context: nil
        )
    }

    private static func squadMapChipLabel(squad: LiveSquadPin, alarming: Bool) -> String {
        let base = (squad.squadName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? squad.squadCode : squad.squadName)
            .prefix(squadChipMaxChars)
            .uppercased()
        return alarming ? "⚠ \(base)" : String(base)
    }

    private static func loadIconBitmap(name: String, targetHeight: CGFloat) -> UIImage {
        guard let image = UIImage(named: name), image.size.height > 0 else {
            return UIImage()
        }
        let scale = targetHeight / image.size.height
        let outW = max(image.size.width * scale, 1)
        let outH = max(targetHeight, 1)
        let renderer = UIGraphicsImageRenderer(size: CGSize(width: outW, height: outH))
        return renderer.image { _ in
            image.draw(in: CGRect(x: 0, y: 0, width: outW, height: outH))
        }
    }
}
